File: backend/services/WaterJugService.cs
using Backend.Models;

namespace Backend.Services
{
    public class WaterJugService : IWaterJugService
    {
        public (List<Step> Steps, bool Solved) SolveWaterJug(int xCap, int yCap, int z)
        {
            if (xCap <= 0 || yCap <= 0 || z <= 0 || z > Math.Max(xCap, yCap))
                return (new List<Step>(), false);

            if (z % Gcd(xCap, yCap) != 0)
                return (new List<Step>(), false);

            var visited = new HashSet<(int, int)>();
            var queue = new Queue<(int x, int y, List<Step> trace)>();
            queue.Enqueue((0, 0, new List<Step>()));

            while (queue.Count > 0)
            {
                var (x, y, trace) = queue.Dequeue();

                if (visited.Contains((x, y))) continue;
                visited.Add((x, y));

                if (x == z || y == z)
                    return (trace, true);

                var nextStates = new List<(int, int, string)>
                {
                    (xCap, y, "Fill bucket X"),
                    (x, yCap, "Fill bucket Y"),
                    (0, y, "Empty bucket X"),
                    (x, 0, "Empty bucket Y")
                };

                int transferXY = Math.Min(x, yCap - y);
                nextStates.Add((x - transferXY, y + transferXY, "Transfer from X to Y"));

                int transferYX = Math.Min(y, xCap - x);
                nextStates.Add((x + transferYX, y - transferYX, "Transfer from Y to X"));

                foreach (var ns in nextStates)
                {
                    var newTrace = new List<Step>(trace); // copiar pasos previos

                    newTrace.Add(new Step
                    {
                        StepNumber = newTrace.Count + 1,
                        BucketX = ns.Item1,
                        BucketY = ns.Item2,
                        Action = ns.Item3
                    });

                    queue.Enqueue((ns.Item1, ns.Item2, newTrace));
                }
            }

            return (new List<Step>(), false);
        }

        private int Gcd(int a, int b)
        {
            while (b != 0)
            {
                int temp = b;
                b = a % b;
                a = temp;
            }
            return a;
        }
    }
}

File: backend/services/waterjug_service.go
package services

import "water-jug-challenge/models"

func SolveWaterJug(xCap, yCap, z int) ([]models.Step, bool) {

	if xCap <= 0 || yCap <= 0 || z <= 0 || z > max(xCap, yCap) {
		return nil, false
	}

	if z%gcd(xCap, yCap) != 0 {
		return nil, false
	}

	type State struct {
		x, y  int
		trace []models.Step
	}

	visited := make(map[[2]int]bool)
	queue := []State{{x: 0, y: 0, trace: []models.Step{}}}

	for len(queue) > 0 {
		curr := queue[0]
		queue = queue[1:]

		if visited[[2]int{curr.x, curr.y}] {
			continue
		}
		visited[[2]int{curr.x, curr.y}] = true

		if curr.x == z || curr.y == z {
			return curr.trace, true
		}

		nextStates := []struct {
			x, y   int
			action string
		}{
			{xCap, curr.y, "Fill bucket X"},
			{curr.x, yCap, "Fill bucket Y"},
			{0, curr.y, "Empty bucket X"},
			{curr.x, 0, "Empty bucket Y"},
		}

		transferXY := min(curr.x, yCap-curr.y)
		nextStates = append(nextStates, struct {
			x, y   int
			action string
		}{curr.x - transferXY, curr.y + transferXY, "Transfer from X to Y"})

		transferYX := min(curr.y, xCap-curr.x)
		nextStates = append(nextStates, struct {
			x, y   int
			action string
		}{curr.x + transferYX, curr.y - transferYX, "Transfer from Y to X"})

		for _, ns := range nextStates {

			newTrace := append([]models.Step{}, curr.trace...)
			newTrace = append(newTrace, models.Step{
				Step:    len(newTrace) + 1,
				BucketX: ns.x,
				BucketY: ns.y,
				Action:  ns.action,
			})
			queue = append(queue, State{x: ns.x, y: ns.y, trace: newTrace})
		}
	}

	return nil, false
}

func gcd(a, b int) int {
	for b != 0 {
		a, b = b, a%b
	}
	return a
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}
